### PR TITLE
execution: refine precision of cast as decimal in agg func

### PR DESCRIPTION
**Pull request by dragonly**
### What problem does this PR solve?

Problem Summary:

Currently the `sum` function will cast [MySQL integer types](https://dev.mysql.com/doc/refman/8.0/en/integer-types.html) `TINYINT, SMALLINT, MEDIUMINT, INT, BIGINT` to decimals with big precision, which is waste of CPU when pushed down to TiFlash.

### What is changed and how it works?

Change the argument decimal precision of `sum` to be just enough for the integer type.

It speeds up TiFlash in a query for TPCH-100 dataset:

```
select sum(L_LINENUMBER) from lineitem;
before: cast int to decimal(32, 0)   1.42s
after: cast int to decimal(10, 0)    1.1s
```

### Benchmark tests

We have conducted benchmark tests with this improvement when the `cast` is executed on TiDB/TiKV/TiFlash.

#### Results

Here's the summary of results. We expect that TiFlash is what really benefits from this improvement, because of it's implementation of different precision of `Decimal`. TiDB/TiKV's results are within acceptable experimental errors.

| engine  | sum(int) |        |        | sum(bigint) |        |        |
|---------|----------|--------|--------|-------------|--------|--------|
|         | before   | after  | %      | before      | after  | %      |
| TiDB    | 28.40s   | 28.79s | +1.4%  | 29.60s      | 28.26s | -4.5%  |
| TiFlash | 1.52s    | 1.18s  | -22.4% | 2.40s       | 1.63s  | -32.1% |
| TiKV    | 29.8s    | 29.09s | -2.4%  | 26.51s      | 26.75s | +0.9%  |

#### plan before PR

These are the plans before this PR respectively:

```sql
-- TiFlash
MySQL [tpch_100]> desc select sum(L_LINENUMBER) from lineitem;
+--------------------------------+--------------+-------------------+----------------+-----------------------------------------------------------------------+
| id                             | estRows      | task              | access object  | operator info                                                         |
+--------------------------------+--------------+-------------------+----------------+-----------------------------------------------------------------------+
| HashAgg_23                     | 1.00         | root              |                | funcs:sum(Column#22)->Column#18                                       |
| └─TableReader_25               | 1.00         | root              |                | data:ExchangeSender_24                                                |
|   └─ExchangeSender_24          | 1.00         | batchCop[tiflash] |                | ExchangeType: PassThrough                                             |
|     └─HashAgg_8                | 1.00         | batchCop[tiflash] |                | funcs:sum(Column#26)->Column#22                                       |
|       └─Projection_37          | 600037902.00 | batchCop[tiflash] |                | cast(tpch_100.lineitem.l_linenumber, decimal(32,0) BINARY)->Column#26 |
|         └─TableFullScan_22     | 600037902.00 | batchCop[tiflash] | table:lineitem | keep order:false                                                      |
+--------------------------------+--------------+-------------------+----------------+-----------------------------------------------------------------------+
6 rows in set (0.00 sec)

MySQL [tpch_100]> desc select sum(cast(L_LINENUMBER as signed)) from lineitem;
+--------------------------------+--------------+-------------------+----------------+------------------------------------------------------------------------------------------------+
| id                             | estRows      | task              | access object  | operator info                                                                                  |
+--------------------------------+--------------+-------------------+----------------+------------------------------------------------------------------------------------------------+
| HashAgg_23                     | 1.00         | root              |                | funcs:sum(Column#22)->Column#18                                                                |
| └─TableReader_25               | 1.00         | root              |                | data:ExchangeSender_24                                                                         |
|   └─ExchangeSender_24          | 1.00         | batchCop[tiflash] |                | ExchangeType: PassThrough                                                                      |
|     └─HashAgg_8                | 1.00         | batchCop[tiflash] |                | funcs:sum(Column#26)->Column#22                                                                |
|       └─Projection_37          | 600037902.00 | batchCop[tiflash] |                | cast(cast(tpch_100.lineitem.l_linenumber, bigint(22) BINARY), decimal(43,0) BINARY)->Column#26 |
|         └─TableFullScan_22     | 600037902.00 | batchCop[tiflash] | table:lineitem | keep order:false                                                                               |
+--------------------------------+--------------+-------------------+----------------+------------------------------------------------------------------------------------------------+
6 rows in set (0.00 sec)

-- TiKV

MySQL [tpch_100]> desc select sum(L_LINENUMBER) from lineitem;
+----------------------------+--------------+-----------+---------------------------------------------------------+------------------------------------------------------+
| id                         | estRows      | task      | access object                                           | operator info                                        |
+----------------------------+--------------+-----------+---------------------------------------------------------+------------------------------------------------------+
| StreamAgg_20               | 1.00         | root      |                                                         | funcs:sum(Column#23)->Column#18                      |
| └─IndexReader_21           | 1.00         | root      |                                                         | index:StreamAgg_8                                    |
|   └─StreamAgg_8            | 1.00         | cop[tikv] |                                                         | funcs:sum(tpch_100.lineitem.l_linenumber)->Column#23 |
|     └─IndexFullScan_19     | 600037902.00 | cop[tikv] | table:lineitem, index:PRIMARY(L_ORDERKEY, L_LINENUMBER) | keep order:false                                     |
+----------------------------+--------------+-----------+---------------------------------------------------------+------------------------------------------------------+

MySQL [tpch_100]> desc select sum(cast(L_LINENUMBER as signed)) from lineitem;
+----------------------------+--------------+-----------+---------------------------------------------------------+-------------------------------------------------------------------------------+
| id                         | estRows      | task      | access object                                           | operator info                                                                 |
+----------------------------+--------------+-----------+---------------------------------------------------------+-------------------------------------------------------------------------------+
| StreamAgg_20               | 1.00         | root      |                                                         | funcs:sum(Column#23)->Column#18                                               |
| └─IndexReader_21           | 1.00         | root      |                                                         | index:StreamAgg_8                                                             |
|   └─StreamAgg_8            | 1.00         | cop[tikv] |                                                         | funcs:sum(cast(tpch_100.lineitem.l_linenumber, bigint(22) BINARY))->Column#23 |
|     └─IndexFullScan_19     | 600037902.00 | cop[tikv] | table:lineitem, index:PRIMARY(L_ORDERKEY, L_LINENUMBER) | keep order:false                                                              |
+----------------------------+--------------+-----------+---------------------------------------------------------+-------------------------------------------------------------------------------+
4 rows in set (0.00 sec)

-- TiDB

MySQL [tpch_100]> set tidb_opt_cpu_factor=0; set tidb_opt_copcpu_factor=100000000000000000;
Query OK, 0 rows affected (0.00 sec)

Query OK, 0 rows affected (0.00 sec)

MySQL [tpch_100]> desc select sum(L_LINENUMBER) from lineitem;
+----------------------------+--------------+-----------+---------------------------------------------------------+-----------------------------------------------------------------------+
| id                         | estRows      | task      | access object                                           | operator info                                                         |
+----------------------------+--------------+-----------+---------------------------------------------------------+-----------------------------------------------------------------------+
| StreamAgg_9                | 1.00         | root      |                                                         | funcs:sum(Column#25)->Column#18                                       |
| └─Projection_26            | 600037902.00 | root      |                                                         | cast(tpch_100.lineitem.l_linenumber, decimal(32,0) BINARY)->Column#25 |
|   └─IndexReader_25         | 600037902.00 | root      |                                                         | index:IndexFullScan_24                                                |
|     └─IndexFullScan_24     | 600037902.00 | cop[tikv] | table:lineitem, index:PRIMARY(L_ORDERKEY, L_LINENUMBER) | keep order:false                                                      |
+----------------------------+--------------+-----------+---------------------------------------------------------+-----------------------------------------------------------------------+
4 rows in set (0.00 sec)

MySQL [tpch_100]> desc select sum(cast(L_LINENUMBER as signed)) from lineitem;
+----------------------------+--------------+-----------+---------------------------------------------------------+------------------------------------------------------------------------------------------------+
| id                         | estRows      | task      | access object                                           | operator info                                                                                  |
+----------------------------+--------------+-----------+---------------------------------------------------------+------------------------------------------------------------------------------------------------+
| StreamAgg_9                | 1.00         | root      |                                                         | funcs:sum(Column#25)->Column#18                                                                |
| └─Projection_26            | 600037902.00 | root      |                                                         | cast(cast(tpch_100.lineitem.l_linenumber, bigint(22) BINARY), decimal(43,0) BINARY)->Column#25 |
|   └─IndexReader_25         | 600037902.00 | root      |                                                         | index:IndexFullScan_24                                                                         |
|     └─IndexFullScan_24     | 600037902.00 | cop[tikv] | table:lineitem, index:PRIMARY(L_ORDERKEY, L_LINENUMBER) | keep order:false                                                                               |
+----------------------------+--------------+-----------+---------------------------------------------------------+------------------------------------------------------------------------------------------------+
4 rows in set (0.00 sec)
```

#### plans after PR

These are the plans after this PR:

```sql
-- TiFlash
MySQL [tpch_100]> desc select sum(L_LINENUMBER) from lineitem;
+--------------------------------+--------------+-------------------+----------------+-----------------------------------------------------------------------+
| id                             | estRows      | task              | access object  | operator info                                                         |
+--------------------------------+--------------+-------------------+----------------+-----------------------------------------------------------------------+
| HashAgg_23                     | 1.00         | root              |                | funcs:sum(Column#22)->Column#18                                       |
| └─TableReader_25               | 1.00         | root              |                | data:ExchangeSender_24                                                |
|   └─ExchangeSender_24          | 1.00         | batchCop[tiflash] |                | ExchangeType: PassThrough                                             |
|     └─HashAgg_8                | 1.00         | batchCop[tiflash] |                | funcs:sum(Column#26)->Column#22                                       |
|       └─Projection_37          | 600037902.00 | batchCop[tiflash] |                | cast(tpch_100.lineitem.l_linenumber, decimal(10,0) BINARY)->Column#26 |
|         └─TableFullScan_22     | 600037902.00 | batchCop[tiflash] | table:lineitem | keep order:false                                                      |
+--------------------------------+--------------+-------------------+----------------+-----------------------------------------------------------------------+
6 rows in set (0.01 sec)

MySQL [tpch_100]> desc select sum(cast(L_LINENUMBER as signed)) from lineitem;
+--------------------------------+--------------+-------------------+----------------+------------------------------------------------------------------------------------------------+
| id                             | estRows      | task              | access object  | operator info                                                                                  |
+--------------------------------+--------------+-------------------+----------------+------------------------------------------------------------------------------------------------+
| HashAgg_23                     | 1.00         | root              |                | funcs:sum(Column#22)->Column#18                                                                |
| └─TableReader_25               | 1.00         | root              |                | data:ExchangeSender_24                                                                         |
|   └─ExchangeSender_24          | 1.00         | batchCop[tiflash] |                | ExchangeType: PassThrough                                                                      |
|     └─HashAgg_8                | 1.00         | batchCop[tiflash] |                | funcs:sum(Column#26)->Column#22                                                                |
|       └─Projection_37          | 600037902.00 | batchCop[tiflash] |                | cast(cast(tpch_100.lineitem.l_linenumber, bigint(22) BINARY), decimal(20,0) BINARY)->Column#26 |
|         └─TableFullScan_22     | 600037902.00 | batchCop[tiflash] | table:lineitem | keep order:false                                                                               |
+--------------------------------+--------------+-------------------+----------------+------------------------------------------------------------------------------------------------+
6 rows in set (0.00 sec)

-- TiKV

MySQL [tpch_100]> desc select sum(L_LINENUMBER) from lineitem;
+----------------------------+--------------+-----------+---------------------------------------------------------+------------------------------------------------------+
| id                         | estRows      | task      | access object                                           | operator info                                        |
+----------------------------+--------------+-----------+---------------------------------------------------------+------------------------------------------------------+
| StreamAgg_20               | 1.00         | root      |                                                         | funcs:sum(Column#23)->Column#18                      |
| └─IndexReader_21           | 1.00         | root      |                                                         | index:StreamAgg_8                                    |
|   └─StreamAgg_8            | 1.00         | cop[tikv] |                                                         | funcs:sum(tpch_100.lineitem.l_linenumber)->Column#23 |
|     └─IndexFullScan_19     | 600037902.00 | cop[tikv] | table:lineitem, index:PRIMARY(L_ORDERKEY, L_LINENUMBER) | keep order:false                                     |
+----------------------------+--------------+-----------+---------------------------------------------------------+------------------------------------------------------+
4 rows in set (0.01 sec)

MySQL [tpch_100]> desc select sum(cast(L_LINENUMBER as signed)) from lineitem;
+----------------------------+--------------+-----------+---------------------------------------------------------+-------------------------------------------------------------------------------+
| id                         | estRows      | task      | access object                                           | operator info                                                                 |
+----------------------------+--------------+-----------+---------------------------------------------------------+-------------------------------------------------------------------------------+
| StreamAgg_20               | 1.00         | root      |                                                         | funcs:sum(Column#23)->Column#18                                               |
| └─IndexReader_21           | 1.00         | root      |                                                         | index:StreamAgg_8                                                             |
|   └─StreamAgg_8            | 1.00         | cop[tikv] |                                                         | funcs:sum(cast(tpch_100.lineitem.l_linenumber, bigint(22) BINARY))->Column#23 |
|     └─IndexFullScan_19     | 600037902.00 | cop[tikv] | table:lineitem, index:PRIMARY(L_ORDERKEY, L_LINENUMBER) | keep order:false                                                              |
+----------------------------+--------------+-----------+---------------------------------------------------------+-------------------------------------------------------------------------------+
4 rows in set (0.00 sec)

-- TiDB

MySQL [tpch_100]> desc select sum(L_LINENUMBER) from lineitem;
+----------------------------+--------------+-----------+---------------------------------------------------------+-----------------------------------------------------------------------+
| id                         | estRows      | task      | access object                                           | operator info                                                         |
+----------------------------+--------------+-----------+---------------------------------------------------------+-----------------------------------------------------------------------+
| StreamAgg_9                | 1.00         | root      |                                                         | funcs:sum(Column#25)->Column#18                                       |
| └─Projection_26            | 600037902.00 | root      |                                                         | cast(tpch_100.lineitem.l_linenumber, decimal(10,0) BINARY)->Column#25 |
|   └─IndexReader_25         | 600037902.00 | root      |                                                         | index:IndexFullScan_24                                                |
|     └─IndexFullScan_24     | 600037902.00 | cop[tikv] | table:lineitem, index:PRIMARY(L_ORDERKEY, L_LINENUMBER) | keep order:false                                                      |
+----------------------------+--------------+-----------+---------------------------------------------------------+-----------------------------------------------------------------------+
4 rows in set (0.00 sec)

MySQL [tpch_100]> desc select sum(cast(L_LINENUMBER as signed)) from lineitem;
+----------------------------+--------------+-----------+---------------------------------------------------------+------------------------------------------------------------------------------------------------+
| id                         | estRows      | task      | access object                                           | operator info                                                                                  |
+----------------------------+--------------+-----------+---------------------------------------------------------+------------------------------------------------------------------------------------------------+
| StreamAgg_9                | 1.00         | root      |                                                         | funcs:sum(Column#25)->Column#18                                                                |
| └─Projection_26            | 600037902.00 | root      |                                                         | cast(cast(tpch_100.lineitem.l_linenumber, bigint(22) BINARY), decimal(20,0) BINARY)->Column#25 |
|   └─IndexReader_25         | 600037902.00 | root      |                                                         | index:IndexFullScan_24                                                                         |
|     └─IndexFullScan_24     | 600037902.00 | cop[tikv] | table:lineitem, index:PRIMARY(L_ORDERKEY, L_LINENUMBER) | keep order:false                                                                               |
+----------------------------+--------------+-----------+---------------------------------------------------------+------------------------------------------------------------------------------------------------+
4 rows in set (0.00 sec)
```

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: expression/aggregation/base_func.go [hunk 1]
```diff
@@ -443,5 +444,20 @@ func (a *baseFuncDesc) WrapCastForAggArgs(ctx sessionctx.Context) {
...
+			case mysql.TypeShort:
+				a.Args[i].GetType().Flen = 5
+			case mysql.TypeInt24:
+				a.Args[i].GetType().Flen = 7
+			case mysql.TypeLong:
+				a.Args[i].GetType().Flen = 10
+			case mysql.TypeLonglong:
+				a.Args[i].GetType().Flen = 19
+			}
+		}
```
Comment (fuzhe1989) [inline expression/aggregation/base_func.go:461]: how about extract them into a function?

File: expression/aggregation/base_func.go [hunk 2]
```diff
@@ -443,5 +444,20 @@ func (a *baseFuncDesc) WrapCastForAggArgs(ctx sessionctx.Context) {
...
+			case mysql.TypeTiny:
+				a.Args[i].GetType().Flen = 3
+			case mysql.TypeShort:
+				a.Args[i].GetType().Flen = 5
+			case mysql.TypeInt24:
+				a.Args[i].GetType().Flen = 7
+			case mysql.TypeLong:
+				a.Args[i].GetType().Flen = 10
+			case mysql.TypeLonglong:
+				a.Args[i].GetType().Flen = 19
```
Comment (fuzhe1989) [inline expression/aggregation/base_func.go:459]: should be 20 if consider unsigned.

File: expression/aggregation/base_func.go [hunk 3]
```diff
@@ -443,5 +444,20 @@ func (a *baseFuncDesc) WrapCastForAggArgs(ctx sessionctx.Context) {
...
 		a.Args[i].GetType().Tp = originTp
+		// refine each mysql integer type to the needed decimal precision
+		if tpOld == mysql.TypeTiny || tpOld == mysql.TypeShort || tpOld == mysql.TypeInt24 || tpOld == mysql.TypeLong || tpOld == mysql.TypeLonglong {
+			switch tpOld {
+			case mysql.TypeTiny:
+				a.Args[i].GetType().Flen = 3
+			case mysql.TypeShort:
+				a.Args[i].GetType().Flen = 5
+			case mysql.TypeInt24:
+				a.Args[i].GetType().Flen = 7
```
Comment (fuzhe1989) [inline expression/aggregation/base_func.go:455]: should be 8 if consider unsigned.

File: expression/aggregation/base_func.go (hunk 1 above)
Comment (dragonly) [reply expression/aggregation/base_func.go]: done

File: expression/aggregation/base_func.go [hunk 4]
```diff
@@ -443,5 +444,31 @@ func (a *baseFuncDesc) WrapCastForAggArgs(ctx sessionctx.Context) {
 		originTp := a.Args[i].GetType().Tp
 		*(a.Args[i].GetType()) = *(a.RetTp)
 		a.Args[i].GetType().Tp = originTp
+		// refine each mysql integer type to the needed decimal precision for sum
+		if a.Name == ast.AggFuncSum && types.IsTypeInteger(tpOld) {
+			if flen, err := refineSumDecimalArgs(tpOld); err != nil {
+				// avg could be split into sum and count, so we should take float decimal into account
+				a.Args[i].GetType().Flen = mathutil.Min(a.Args[i].GetType().Flen, flen+a.Args[i].GetType().Decimal)
```
Comment (fuzhe1989) [inline expression/aggregation/base_func.go:451]: also explain the `mathutil.Min`?

File: expression/aggregation/base_func.go [hunk 5]
```diff
@@ -443,5 +444,31 @@ func (a *baseFuncDesc) WrapCastForAggArgs(ctx sessionctx.Context) {
...
+func refineSumDecimalArgs(tp byte) (int, error) {
+	switch tp {
+	case mysql.TypeTiny:
+		return 3, nil
+	case mysql.TypeShort:
+		return 5, nil
+	case mysql.TypeInt24:
+		return 8, nil
+	case mysql.TypeLong:
+		return 1, nil
```
Comment (fuzhe1989) [inline expression/aggregation/base_func.go:466]:
```suggestion
		return 10, nil
```

File: expression/aggregation/base_func.go [hunk 6]
```diff
@@ -443,5 +444,31 @@ func (a *baseFuncDesc) WrapCastForAggArgs(ctx sessionctx.Context) {
...
+	case mysql.TypeLonglong:
+		return 2, nil
```
Comment (fuzhe1989) [inline expression/aggregation/base_func.go:468]:
```suggestion
		return 20, nil
```

File: expression/aggregation/base_func.go (hunk 4 above)
Comment (dragonly) [reply expression/aggregation/base_func.go:451]: just integer min.
it's already in this file before this PR.

File: expression/aggregation/base_func.go [hunk 7]
```diff
@@ -443,5 +444,31 @@ func (a *baseFuncDesc) WrapCastForAggArgs(ctx sessionctx.Context) {
...
+		if a.Name == ast.AggFuncSum && types.IsTypeInteger(tpOld) {
+			if flen, err := refineSumDecimalArgs(tpOld); err != nil {
+				// avg could be split into sum and count, so we should take float decimal into account
+				a.Args[i].GetType().Flen = mathutil.Min(a.Args[i].GetType().Flen, flen+a.Args[i].GetType().Decimal)
+			}
+		}
+	}
+}
+
+func refineSumDecimalArgs(tp byte) (int, error) {
```
Comment (fuzhe1989) [inline expression/aggregation/base_func.go:457]:
```suggestion
func minimalDecimalLenForHoldingInteger(tp byte) (int, error) {
```

File: expression/aggregation/base_func.go [hunk 8]
```diff
@@ -443,5 +444,31 @@ func (a *baseFuncDesc) WrapCastForAggArgs(ctx sessionctx.Context) {
...
+	case mysql.TypeInt24:
+		return 8, nil
+	case mysql.TypeLong:
+		return 10, nil
+	case mysql.TypeLonglong:
+		return 20, nil
+	case mysql.TypeYear:
+		return 4, nil
+	default:
+		return -1, errors.New("")
```
Comment (XuHuaiyu) [inline expression/aggregation/base_func.go:472]: Add an error message

File: expression/aggregation/base_func.go [hunk 9]
```diff
@@ -443,5 +444,31 @@ func (a *baseFuncDesc) WrapCastForAggArgs(ctx sessionctx.Context) {
 		originTp := a.Args[i].GetType().Tp
 		*(a.Args[i].GetType()) = *(a.RetTp)
 		a.Args[i].GetType().Tp = originTp
+		// refine each mysql integer type to the needed decimal precision for sum
+		if a.Name == ast.AggFuncSum && types.IsTypeInteger(tpOld) {
+			if flen, err := minimalDecimalLenForHoldingInteger(tpOld); err != nil {
+				// avg could be split into sum and count, so we should take float decimal into account
```
Comment (XuHuaiyu) [inline expression/aggregation/base_func.go:450]: float decimal --> \`Decimal\` is more readable?

File: expression/aggregation/base_func.go (hunk 8 above)
Comment (dragonly) [reply expression/aggregation/base_func.go]: done

File: expression/aggregation/base_func.go (hunk 9 above)
Comment (dragonly) [reply expression/aggregation/base_func.go]: done

**Review by XuHuaiyu — APPROVED** [review]
LGTM

File: expression/aggregation/base_func.go [hunk 10]
```diff
@@ -443,5 +444,31 @@ func (a *baseFuncDesc) WrapCastForAggArgs(ctx sessionctx.Context) {
 		originTp := a.Args[i].GetType().Tp
 		*(a.Args[i].GetType()) = *(a.RetTp)
 		a.Args[i].GetType().Tp = originTp
+		// refine each mysql integer type to the needed decimal precision for sum
```
Comment (fuzhe1989) [inline expression/aggregation/base_func.go:447]: Besides `minimalDecimalLenForHoldingInteger`, how about extract L447-L454 to another function? `adjustDecimalLenForSumInteger`?

File: expression/aggregation/base_func.go [hunk 11]
```diff
@@ -443,5 +444,31 @@ func (a *baseFuncDesc) WrapCastForAggArgs(ctx sessionctx.Context) {
...
+		if a.Name == ast.AggFuncSum && types.IsTypeInteger(tpOld) {
+			if flen, err := minimalDecimalLenForHoldingInteger(tpOld); err != nil {
+				// avg could be split into sum and count, so we should take the `.Decimal` field into account
+				a.Args[i].GetType().Flen = mathutil.Min(a.Args[i].GetType().Flen, flen+a.Args[i].GetType().Decimal)
```
Comment (fuzhe1989) [inline expression/aggregation/base_func.go:451]: nit: append comment like "use mathutil.Min here to ensure flen won't increase"?

File: expression/aggregation/base_func.go [hunk 12]
```diff
@@ -443,5 +444,31 @@ func (a *baseFuncDesc) WrapCastForAggArgs(ctx sessionctx.Context) {
 		originTp := a.Args[i].GetType().Tp
 		*(a.Args[i].GetType()) = *(a.RetTp)
 		a.Args[i].GetType().Tp = originTp
+		// refine each mysql integer type to the needed decimal precision for sum
+		if a.Name == ast.AggFuncSum && types.IsTypeInteger(tpOld) {
```
Comment (fuzhe1989) [inline expression/aggregation/base_func.go:448]: do we need to check the new type is decimal? @XuHuaiyu

**Review by guo-shaoge — APPROVED** [review]

File: expression/aggregation/base_func.go (hunk 10 above)
Comment (dragonly) [reply expression/aggregation/base_func.go:447]: This function is so tightly coupled with the surrounding code (like the iterating index `i`, and the if-true-then-modify logic), which results in many input/output of the extracted function (not so clean abstraction).

File: expression/aggregation/base_func.go (hunk 11 above)
Comment (dragonly) [reply expression/aggregation/base_func.go:451]: done

File: expression/aggregation/base_func.go (hunk 10 above)
Comment (fuzhe1989) [reply expression/aggregation/base_func.go:448]:
```go
func adjustDecimalLenForSumInteger(ft *FieldType, tpOld byte) {
    if types.IsTypeInteger(tpOld) && ft.Tp == mysql.TypeNewDecimal {
        if flen, err := minimalDecimalLenForHoldingInteger(tpOld); err == nil {
            ft.Flen = mathutil.Min(ft.Flen, flen+ft.Decimal)
        }
    }
}
```

Comment (fuzhe1989) [reply expression/aggregation/base_func.go:448]: then here is:
```go
if a.Name == ast.AggFuncSum {
    adjustDecimalLenForSumInteger(a.Args[i].GetType(), tpOld)
}
```

An explicit function can be more readable for readers who don't care the details.

File: expression/aggregation/base_func.go [hunk 13]
```diff
@@ -443,5 +444,33 @@ func (a *baseFuncDesc) WrapCastForAggArgs(ctx sessionctx.Context) {
 		originTp := a.Args[i].GetType().Tp
 		*(a.Args[i].GetType()) = *(a.RetTp)
 		a.Args[i].GetType().Tp = originTp
+
+		// refine each mysql integer type to the needed decimal precision for sum
+		if a.Name == ast.AggFuncSum && types.IsTypeInteger(tpOld) {
+			if flen, err := minimalDecimalLenForHoldingInteger(tpOld); err != nil {
```
Comment (fuzhe1989) [inline expression/aggregation/base_func.go:450]:
```suggestion
			if flen, err := minimalDecimalLenForHoldingInteger(tpOld); err == nil {
```

Comment (dragonly) [reply expression/aggregation/base_func.go:450]: OMG...

File: expression/aggregation/base_func.go (hunk 12 above)
Comment (dragonly) [reply expression/aggregation/base_func.go:449]: it's done in the function extraction refactor above.

**Comment by siddontang** [comment]
@zanmato1984 

why TiKV can't get too much benefit from this change?
another thing is I only see the integration tests, no any unit tests?

**Comment by fuzhe1989** [comment]
@siddontang TiDB will push the whole agg down to TiKV without appending any cast. This will also work for TiFlash in the future.

About unittest, this pr doesn't change any operator or function, the only thing it affects is the generated physical plan and I think explaintest is enough.

**Pull request merged**